share/fsmon/fanotify_linux.go

```diff
@@ -560,53 +563,62 @@ func (fn *FaNotify) MonitorFileEvents() {
-
-			resp, mask, nvPod, ifile, pInfo := fn.calculateResponse(pid, fd, fmask, perm)
-			if perm {
-				if err := fn.fa.Response(ev, resp); err != nil {
-					log.WithFields(log.Fields{"err": err, "pid": pid}).Error()
-				}
+	for {
+		ev, err := fn.fa.GetEvent(os.Getpid())
+		if err != nil {
+			return err
+		}
+		if ev == nil {
+			return nil
+		}
+		pid := int(ev.Pid)
+		fd := int(ev.Fd)
+		fmask := uint64(ev.Mask)
+		perm := (fmask & (unix.FAN_OPEN_PERM | unix.FAN_ACCESS_PERM)) > 0
+
+		// log.WithFields(log.Fields{"pid": pid, "fmask": fmt.Sprintf("0x%08x", fmask), "fd": fd}).Debug("FMON:")
+		resp, mask, nvPod, ifile, pInfo := fn.calculateResponse(pid, fd, fmask, perm)
+		if perm {
+			if resp {
+				err = fn.fa.ResponseAllow(ev)
+			} else {
+				err = fn.fa.ResponseDeny(ev)
 			}
-			ev.File.Close()
-
-			if nvPod {
-				if !resp && ifile != nil && pInfo != nil {
-					finfo := ifile.params.(*osutil.FileInfoExt)
-					_, path := fn.sys.ParseContainerFilePath(ifile.path)
-					log.WithFields(log.Fields{"path": path, "caller": pInfo.Path, "pid": pid}).Info("FMON: NV Protect")
-					go fn.sendNVrpt(pInfo.RootPid, pid, finfo.ContainerId, path, pInfo.Path)
-				}
-				continue
+			if err != nil {
+				log.WithFields(log.Fields{"err": err, "pid": pid, "resp": resp}).Error()
 			}
+		}
+		ev.Close()
 
-			if ifile == nil {
-				continue // nothing to justify
+		if nvPod {
+			if !resp && ifile != nil && pInfo != nil {
+				finfo := ifile.params.(*osutil.FileInfoExt)
+				_, path := fn.sys.ParseContainerFilePath(ifile.path)
+				log.WithFields(log.Fields{"path": path, "caller": pInfo.Path, "pid": pid}).Info("FMON: NV Protect")
+				go fn.sendNVrpt(pInfo.RootPid, pid, finfo.ContainerId, path, pInfo.Path)
 			}
+			continue
+		}
 
-			change := (fmask & FAN_CLOSE_WRITE) > 0
-			// log.WithFields(log.Fields{"ifile": ifile, "pInfo": pInfo, "Resp": resp, "Change": change, "Perm": perm}).Debug("FMON:")
+		if ifile == nil {
+			continue // nothing to justify
+		}
 
-			var bReporting bool
-			if ifile.learnt { // discover mode
-				bReporting = ifile.userAdd // learn app for customer-added entry
-			} else { // monitor or protect mode
-				allowRead := resp && !change
-				bReporting = !allowRead // allowed app by block_access
-			}
+		change := (fmask & unix.FAN_CLOSE_WRITE) > 0
+		// log.WithFields(log.Fields{"ifile": ifile, "pInfo": pInfo, "Resp": resp, "Change": change, "Perm": perm}).Debug("FMON:")
 
-			if bReporting || change { // report changed file
-				ifile.cb(ifile.path, mask, ifile.params, pInfo)
-			}
+		var bReporting bool
+		if ifile.learnt { // discover mode
+			bReporting = ifile.userAdd // learn app for customer-added entry
+		} else { // monitor or protect mode
+			allowRead := resp && !change
+			bReporting = !allowRead // allowed app by block_access
+		}
+
+		if bReporting || change { // report changed file
+			ifile.cb(ifile.path, mask, ifile.params, pInfo)
 		}
 	}
-	return nil
```

Comment:
Hmm, why are we removing this line?

Reply:
My local lint complaints that it was unreachable "return" in this function.